usr/sbin/p11sak/p11tool.h

```diff
@@ -0,0 +1,401 @@
+
+enum p11tool_arg_type {
+    ARG_TYPE_PLAIN = 0, /* no argument */
+    ARG_TYPE_STRING = 1,
+    ARG_TYPE_ENUM = 2,
+    ARG_TYPE_NUMBER = 3,
+};
+
+struct p11tool_enum_value {
+    const char *value;
+    const struct p11tool_arg *args;
+    union {
+        const void *ptr;
+        CK_ULONG num;
+    } private;
+    char **any_value; /* if this is not NULL then this enum value matches to
+                         any string, and the string is set into any_value */
+};
+
+struct p11tool_arg {
+    const char *name;
+    enum p11tool_arg_type type;
+    bool required;
+    bool case_sensitive;
+    const struct p11tool_enum_value *enum_values;
+    union {
+        bool *plain;
+        char **string;
+        struct p11tool_enum_value **enum_value;
+        CK_ULONG *number;
+    } value;
+    bool (*is_set)(const struct p11tool_arg *arg);
+    const char *description;
+};
+
+struct p11tool_opt {
+    char short_opt; /* 0 if no short option is used */
+    const char *long_opt; /* NULL if no long option */
+    int long_opt_val; /* Used only if short_opt is 0 */
+    bool required;
+    struct p11tool_arg arg;
+    const char *description;
+};
+
+struct p11tool_cmd {
+    const char *cmd;
+    const char *cmd_short1;
+    const char *cmd_short2;
+    CK_RV (*func)(void);
+    const struct p11tool_opt *opts;
+    const struct p11tool_arg *args;
+    const char *description;
+    void (*help)(void);
+    CK_FLAGS session_flags;
+};
+
+struct p11tool_attr {
+    const char *name;
+    CK_ATTRIBUTE_TYPE type;
+    char letter;
+    bool secret;
+    bool public;
+    bool private;
+    bool settable;
+    bool so_set_to_true; /* can only be set to TRUE by SO */
+    void (*print_short)(const CK_ATTRIBUTE *val, bool applicable);
+    void (*print_long)(const char *attr, const CK_ATTRIBUTE *val,
+                       int indent, bool sensitive);
+};
+
+#define DECLARE_BOOL_ATTR(attr, ch, sec, pub, priv, set)                       \
+    { .name = # attr, .type = attr, .letter = ch,                              \
+      .secret = sec, .public = pub, .private = priv,                           \
+      .settable = set, .print_short = p11tool_print_bool_attr_short,           \
+      .print_long = p11tool_print_bool_attr_long, }
+
+#define DECLARE_BOOL_ATTR_SO(attr, ch, sec, pub, priv, set, so_set_true)       \
+    { .name = # attr, .type = attr, .letter = ch,                              \
+      .secret = sec, .public = pub, .private = priv,                           \
+      .settable = set, .so_set_to_true = so_set_true,                          \
+      .print_short = p11tool_print_bool_attr_short,                            \
+      .print_long = p11tool_print_bool_attr_long, }
+
+struct p11tool_objtype {
+    const char *obj_typestr;
+    const char *obj_liststr;
+    const char *name;
+    CK_ULONG type; /* CK_KEY_TYPE or CK_CERTIFICATE_TYPE */
+    const char *ck_name;
+    int pkey_type; /* OpenSSL PKEY_xxx type or 0 if not applicable */
```

Comment:
pkey.h uses type u32 for the pkey_type. Maybe unsigned int here?

Reply:
This is the OpenSSL key type (they also call an asym key 'PKEY'), this is not at all related to the pkey kernel module. So this must stay an int.